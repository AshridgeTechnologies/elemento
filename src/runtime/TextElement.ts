import React from 'react'

export function TextElement(props: {children: any}) {
    const {children} = props
    return React.createElement('p', null, children)
}
