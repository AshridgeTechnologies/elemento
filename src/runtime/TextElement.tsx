import React from 'react'
import {Typography} from '@mui/material'
import {isArray} from 'lodash'
import {asText, valueOfProps} from './runtimeFunctions'

type Properties = {path: string, children?: any
    fontSize?: string | number, fontFamily?: string, color?: string, backgroundColor?: string,
    border?: number, borderColor?: string, width?: number, height?: number
    }

export default function TextElement({children, path, ...props}: Properties) {
    const childArray = isArray(children) ? children : [children]
    const propVals = valueOfProps(props)
    return React.createElement(Typography, {id: path, gutterBottom: true, ...propVals}, ...childArray.map(asText))
}
