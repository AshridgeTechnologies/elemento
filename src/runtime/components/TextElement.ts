import React from 'react'
import {Typography} from '@mui/material'
import lodash from 'lodash'; const {isArray, isFunction, isPlainObject, isObject, isString} = lodash;
import {PropVal, StylesProps, valueLiteral, valueOfProps} from '../runtimeFunctions'
import {flatten} from 'ramda'

type Properties = Readonly<{path: string, children?: any,
    show?: PropVal<boolean>,
    styles?: StylesProps
    }>

function asText(content: any) {
    if (React.isValidElement(content)) return content
    const contentValue = content?.valueOf()
    if (isString(contentValue) && contentValue.includes('\n')) {
        return flatten(contentValue.split('\n').map((line: string, index: number) => [line, React.createElement('br', {key: index})]))
    }
    if (isPlainObject(contentValue)) {
        return valueLiteral(contentValue)
    }
    if (isFunction(contentValue)) {
        return 'function ' + contentValue.name
    }
    if (isObject(contentValue)) {
        return contentValue.toString()
    }
    return contentValue
}

export default function TextElement({children, path, ...props}: Properties) {
    const childArray = isArray(children) ? children : [children]
    const {show = true, styles = {}} = valueOfProps(props)
    const showProps = show ? {} : {display: 'none'}
    const sxProps = {
        ...showProps,
        ...styles
    }

    return React.createElement(Typography, {id: path, sx: sxProps}, ...childArray.map(asText))
}
