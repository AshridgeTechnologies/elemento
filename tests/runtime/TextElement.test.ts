import React from 'react'
import renderer from 'react-test-renderer'
import {TextElement} from '../../src/runtime/TextElement.js'

test('TextElement element produces output containing children', ()=> {
    const component = renderer.create(React.createElement(TextElement, null, 'Hello', 'where are you'))
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})