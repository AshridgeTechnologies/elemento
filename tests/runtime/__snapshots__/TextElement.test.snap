// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextElement element produces output containing children 1`] = `
<p>
  Hello
  where are you
</p>
`;
